web/models/requirement.ex

```diff
@@ -4,120 +4,43 @@ defmodule HexWeb.Requirement do
+  def changeset(requirement, params \\ %{}) do
+    changeset = cast(requirement, params, ~w(name app requirement optional))
 
-  def create_all(release, requirements) do
-    requirements = normalize(requirements)
-    deps = deps(requirements)
+    name = changeset.changes.name
+    app = Map.get(changeset.changes, :app, name)
+    dep = HexWeb.Repo.get_by!(Package, name: name)
 
-    errors = Enum.map(requirements, &validate(deps, &1))
-             |> Enum.filter(&match?({:error, _}, &1))
-
-    if errors == [] do
-      case resolve(requirements, guess_config(release)) do
-        :ok ->
-          Enum.each(requirements, &insert(release, deps, &1))
-          {:ok, requirements}
-        {:error, reason} ->
-          {:error, reason}
-      end
-    else
-      {:error, Enum.map(errors, &elem(&1, 1))}
-    end
-  end
-
-  defp deps(requirements) do
-    deps = Enum.map(requirements, & &1.name)
-
-    from(p in Package,
-         where: p.name in ^deps,
-         select: {p.name, p.id})
-    |> HexWeb.Repo.all
-    |> Enum.into(%{})
+    changeset
+    |> put_change(:app, app)
+    |> put_change(:dependency_id, dep.id)
+    |> validate_requirement
   end
 
-  defp normalize(requirements) do
-    Enum.map(requirements, fn
-      {dep, map} when is_map(map) ->
-        %{name: to_string(dep), app: map["app"], requirement: map["requirement"], optional: map["optional"] || false}
-      {dep, {req, app}} ->
-        %{name: to_string(dep), app: to_string(app), requirement: req, optional: false}
-      {dep, req} ->
-        %{name: to_string(dep), app: to_string(dep), requirement: req, optional: false}
+  defp validate_requirement(changeset) do
+    validate_change(changeset, :requirement, fn key, req ->
+      cond do
+        is_nil(req) ->
+          # Temporary friendly error message until people update to hex 0.9.1
+          [{key, "invalid requirement: #{inspect req}, use \">= 0.0.0\" instead"}]
+        not valid?(req) ->
+          [{key, "invalid requirement: #{inspect req}"}]
+        true ->
+          []
+      end
     end)
   end
 
-  defp validate(deps, %{name: dep, requirement: req}) do
-    cond do
-      is_nil(req) ->
-        # Temporary friendly error message until people update to hex 0.9.1
-        {:error, {dep, "invalid requirement: #{inspect req}, use \">= 0.0.0\" instead"}}
-      not valid?(req) ->
-        {:error, {dep, "invalid requirement: #{inspect req}"}}
-      !deps[dep] ->
-        {:error, {dep, "unknown package"}}
-      true ->
-        :ok
-    end
-  end
-
-  defp insert(release, deps, %{name: dep, app: app, requirement: req, optional: optional}) do
-    build_assoc(release, :requirements)
-    |> struct(requirement: req,
-              app: app,
-              optional: optional,
-              dependency_id: deps[dep])
-    |> HexWeb.Repo.insert!
-  end
-
   defp valid?(req) do
     is_binary(req) and match?({:ok, _}, Version.parse_requirement(req))
   end
-
-  defp resolve(requirements, config) do
```

Comment:
Yeah that makes sense. I ended up with a similar structure as on master, there's again the  `Requirement.create_all` function which now works on `release_changeset` (instead of `release`) and it doesn't write anything to the DB anymore. See: 344cd54be3de28b3212cba3d0a786102442350c6

It still needs some work, as there's unnecessary 1 DB call per requirement and the logic around checking errors for conflicts got really complicated. The main issue I think is going back & forth between the way we need to display requirement errors on releases endpoint (to keep backwards compatibility with clients) as well as having some errors (e.g. invalid requirement string) on a particular package as well some to the requirements as a whole for the conflicts resolution (for conflict I ended up annotating every requirement with the same error). Anyhow, I'll keep at it, there's definitely some ways to clean it up even further!